util/eventbus/eventbustest/eventbustest.go

```diff
@@ -0,0 +1,229 @@
+	// of an event, the TimeOut can be set to a lower value after creating the
+	// Watcher.
+	TimeOut time.Duration
+}
+
+// Expect is a particular implementation of [ExpectFunc] that tests for the
+// existence of an event without caring about the contents of it.
+func Expect[T any](tw *Watcher) error {
+	return ExpectFunc(tw, func(event T) (bool, error) { return true, nil })
+}
+
+// ExpectFunc tests for a particular event but also a particular shape of said event.
+// This allows for looking for a single event with a specific value, or a
+// particular event in a pile of SomeEvent.
+func ExpectFunc[T any](tw *Watcher, test func(event T) (bool, error)) error {
+	eventCount := 0
+	for {
+		select {
+		case event := <-tw.events:
+			eventCount = eventCount + 1
+			if ev, ok := event.(T); ok {
+				if ok, err := test(ev); ok {
+					return nil
+				} else if err != nil {
+					return err
+				}
+			}
+		case <-time.After(tw.TimeOut):
+			return fmt.Errorf("timed out waiting for event, saw %d events", eventCount)
+		}
+	}
+}
+
+// ExpectEvents tests for a particular set of events in an order but other events
+// can happen at the same time. The contents of the events will not be tested,
+// only the event type itself.
+func ExpectEvents(tw *Watcher, events ...any) error {
+	eventCount := 0
+	head := 0
+	for head < len(events) {
+		select {
+		case event := <-tw.events:
+			eventCount = eventCount + 1
+			typeEvent := reflect.TypeOf(event)
+			typeHead := reflect.TypeOf(events[head])
+			if typeEvent == typeHead {
+				head = head + 1
+			}
+		case <-time.After(tw.TimeOut):
+			return fmt.Errorf(
+				"timed out waiting for event, saw %d events, %d was expected",
+				eventCount, head)
+		}
+	}
+	return nil
+}
+
+// ExpectOnlyEvents tests for a particular set of events in an order but other events
+// are not allowed to happen at the same time. The contents of the events will not be tested,
+// only the event type itself.
+func ExpectOnlyEvents(tw *Watcher, events ...any) error {
+	eventCount := 0
+	head := 0
+	for head < len(events) {
+		select {
+		case event := <-tw.events:
+			eventCount = eventCount + 1
+			typeEvent := reflect.TypeOf(event)
+			typeHead := reflect.TypeOf(events[head])
+			if typeEvent != typeHead {
+				return fmt.Errorf(
+					"expected event type %s, saw %s, at index %d",
+					typeHead, typeEvent, head)
+			}
+			head = head + 1
+		case <-time.After(tw.TimeOut):
+			return fmt.Errorf(
+				"timed out waiting for event, saw %d events, %d was expected",
+				eventCount, head)
+		}
+	}
+	return nil
+}
+
+// EventFunc is a test helper type that holds an event an a function to test the
+// validity of that event. F returns indicates:
+//
+//	true, nil - Event matches, continue processing
+//	false, nil - Event does not match, skip and continue processing
+//	*, error - Stop processing, return error
+//
+// See [ExpectEventsFunc] for example.
+type EventFunc struct {
+	Event any
```

Comment:
```suggestion
	// Event identifies the event type to be matched. It must be non-nil.
	Event any
```

this makes me wonder, though, should we let this be nil and then pass _any_ event type to the predicate?

Reply:
I had a crazy idea. What if instead of a struct, we allow the caller to pass in any function of the general shape

    func(T) bool
    func(T) (bool, error)

and then use reflection to pull out the types? It's gross, but the grossness is hidden and the reflection cost doesn't matter in a test.

Very rough prototype: https://go.dev/play/p/HW_moR7aj3F


So now you could say:

```go
ExpectEvents(tw, func(Thing) bool { return true }, func(o Other) bool { return o.Value == 5 }, func(o Other) (bool, error) { ... })
```

and it would figure out the argument type from the function signature.

This may be too crazy to live, so I am not asking you to do this, but I would be curious for your thoughts.

Reply:
Lol, but it does raise an interesting point since we can consolidate some functions.

Pros are:
 - Fewer functions to know and understand
 - Fewer functions to maintain

Cons are:
 - The simple use case of only checking types gets more complicated to use (IMO)
 - Maintaining the reflect code might be slightly more expensive

I do though think that I like this better than having a struct to pass in. So I think I will move the current `ExpectEventsFunc` to an approach like this.

Reply:
Yeah, it probably would lstill make sense to have a wrapper that takes "type only" arguments like the `Events` helper does now. Or perhaps we provide a helper, e.g.,

```go
func Type[T any]() func(T) bool { return func(T) bool { return true } }
```

as syntactic sugar

```go
eventbustest.Events(tw, eventbustest.Type[E1](), eventbustest.Type[E2](), ...)
```

maybe